File: src/memory/block.ts
// Block model for memory files (DESIGN §5 "Block model").
//
// Every memory file may carry YAML-ish frontmatter with `description` (always visible to the
// manager even when the body isn't) and an optional `limit` (character budget). We hand-roll a
// tiny parser rather than pull in a YAML dependency — the design's "zero external deps for
// storage" rule. Only flat `key: value` scalars are supported; that is all the block model needs.

export interface MemoryBlock {
  /** One-line summary shown in the archival tree even when the body is not loaded. */
  description?: string;
  /** Optional character budget for the body (advisory; surfaced to the manager). */
  limit?: number;
  /** Everything after the frontmatter (the actual content). */
  body: string;
}

const FENCE = "---";

/** Parse a raw file into { description, limit, body }. No frontmatter → body is the whole file. */
export function parseBlock(raw: string): MemoryBlock {
  const normalized = raw.replace(/\r\n/g, "\n");
  if (!normalized.startsWith(FENCE + "\n")) {
    return { body: normalized };
  }
  const end = normalized.indexOf("\n" + FENCE, FENCE.length);
  if (end === -1) {
    // Unterminated frontmatter — treat the whole thing as body rather than throwing.
    return { body: normalized };
  }
  const frontmatter = normalized.slice(FENCE.length + 1, end);
  // Body starts after the closing fence line.
  const afterFence = normalized.indexOf("\n", end + 1);
  const body = afterFence === -1 ? "" : normalized.slice(afterFence + 1);

  const block: MemoryBlock = { body };
  for (const line of frontmatter.split("\n")) {
    const m = /^([A-Za-z_][\w-]*)\s*:\s*(.*)$/.exec(line.trim());
    if (!m) continue;
    const key = m[1]!;
    const value = stripQuotes(m[2]!.trim());
    if (key === "description") block.description = value;
    else if (key === "limit") {
      const n = Number(value);
      if (Number.isFinite(n)) block.limit = n;
    }
  }
  return block;
}

/** Serialize a block back to disk, emitting frontmatter only when there is metadata to record. */
export function serializeBlock(block: MemoryBlock): string {
  const hasMeta = block.description !== undefined || block.limit !== undefined;
  if (!hasMeta) return block.body;
  const lines = [FENCE];
  if (block.description !== undefined) lines.push(`description: ${block.description}`);
  if (block.limit !== undefined) lines.push(`limit: ${block.limit}`);
  lines.push(FENCE, "");
  return lines.join("\n") + block.body;
}

/** The text we index for search: description + body, so a file is findable by either. */
export function indexableText(block: MemoryBlock): string {
  return [block.description ?? "", block.body].filter(Boolean).join("\n");
}

function stripQuotes(s: string): string {
  if (s.length >= 2 && ((s.startsWith('"') && s.endsWith('"')) || (s.startsWith("'") && s.endsWith("'")))) {
    return s.slice(1, -1);
  }
  return s;
}

File: src/memory/block.rs
//! Block model for memory files. Port of `src/memory/block.ts`.
//!
//! Every memory file may carry frontmatter with `description` (always visible to the manager even
//! when the body isn't) and an optional `limit`. We hand-roll a tiny parser rather than pull in a
//! YAML dependency — only flat `key: value` scalars are supported.

const FENCE: &str = "---";

/// Parsed memory file: optional metadata + body.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct MemoryBlock {
    /// One-line summary shown in the tree even when the body is not loaded.
    pub description: Option<String>,
    /// Optional character budget for the body (advisory).
    pub limit: Option<i64>,
    /// Everything after the frontmatter.
    pub body: String,
}

/// Parse a raw file into a [`MemoryBlock`]. No (or unterminated) frontmatter → whole file is body.
pub fn parse_block(raw: &str) -> MemoryBlock {
    let normalized = raw.replace("\r\n", "\n");
    let fence_open = format!("{FENCE}\n");
    if !normalized.starts_with(&fence_open) {
        return MemoryBlock {
            body: normalized,
            ..Default::default()
        };
    }
    // Find the closing fence ("\n---") after the opening one.
    let Some(end) = normalized[FENCE.len()..]
        .find(&format!("\n{FENCE}"))
        .map(|i| i + FENCE.len())
    else {
        return MemoryBlock {
            body: normalized,
            ..Default::default()
        };
    };
    let frontmatter = &normalized[FENCE.len() + 1..end];
    // Body starts after the line containing the closing fence.
    let body = match normalized[end + 1..].find('\n') {
        Some(nl) => normalized[end + 1 + nl + 1..].to_string(),
        None => String::new(),
    };

    let mut block = MemoryBlock {
        body,
        ..Default::default()
    };
    for line in frontmatter.split('\n') {
        if let Some((key, value)) = parse_frontmatter_line(line.trim()) {
            match key.as_str() {
                "description" => block.description = Some(value),
                "limit" => block.limit = value.parse::<i64>().ok(),
                _ => {}
            }
        }
    }
    block
}

/// Parse a `key: value` frontmatter line (key starts with a letter/underscore). Returns the key and
/// the de-quoted value.
fn parse_frontmatter_line(line: &str) -> Option<(String, String)> {
    let (key, value) = line.split_once(':')?;
    let key = key.trim();
    if key.is_empty()
        || !key
            .chars()
            .next()
            .is_some_and(|c| c.is_ascii_alphabetic() || c == '_')
    {
        return None;
    }
    if !key
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
    {
        return None;
    }
    Some((key.to_string(), strip_quotes(value.trim()).to_string()))
}

/// Serialize a block to disk, emitting frontmatter only when there is metadata to record.
pub fn serialize_block(block: &MemoryBlock) -> String {
    if block.description.is_none() && block.limit.is_none() {
        return block.body.clone();
    }
    let mut out = String::from(FENCE);
    out.push('\n');
    if let Some(desc) = &block.description {
        out.push_str(&format!("description: {desc}\n"));
    }
    if let Some(limit) = block.limit {
        out.push_str(&format!("limit: {limit}\n"));
    }
    out.push_str(FENCE);
    out.push('\n');
    out.push_str(&block.body);
    out
}

/// The text we index for search: description + body, so a file is findable by either.
pub fn indexable_text(block: &MemoryBlock) -> String {
    let mut parts = Vec::new();
    if let Some(desc) = &block.description
        && !desc.is_empty()
    {
        parts.push(desc.clone());
    }
    if !block.body.is_empty() {
        parts.push(block.body.clone());
    }
    parts.join("\n")
}

fn strip_quotes(s: &str) -> &str {
    let bytes = s.as_bytes();
    if bytes.len() >= 2 {
        let (first, last) = (bytes[0], bytes[bytes.len() - 1]);
        if (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'') {
            return &s[1..s.len() - 1];
        }
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_frontmatter() {
        let b = parse_block("---\ndescription: hi there\nlimit: 200\n---\nbody line\n");
        assert_eq!(b.description.as_deref(), Some("hi there"));
        assert_eq!(b.limit, Some(200));
        assert_eq!(b.body, "body line\n");
    }

    #[test]
    fn no_frontmatter_is_all_body() {
        let b = parse_block("just a body\nsecond line");
        assert!(b.description.is_none());
        assert_eq!(b.body, "just a body\nsecond line");
    }

    #[test]
    fn roundtrips() {
        let b = MemoryBlock {
            description: Some("d".into()),
            limit: None,
            body: "the body\n".into(),
        };
        let s = serialize_block(&b);
        assert_eq!(parse_block(&s), b);
    }

    #[test]
    fn indexable_includes_description_and_body() {
        let b = parse_block("---\ndescription: findme\n---\nand the body too\n");
        let idx = indexable_text(&b);
        assert!(idx.contains("findme") && idx.contains("body"));
    }
}
